atlasdb-config/src/main/java/com/palantir/atlasdb/factory/timelock/TimestampCorroboratingTimelockService.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.factory.timelock;
+
+import java.util.function.Supplier;
+import java.util.function.ToLongFunction;
+
+import com.palantir.lock.v2.AutoDelegate_TimelockService;
+import com.palantir.lock.v2.IdentifiedTimeLockRequest;
+import com.palantir.lock.v2.StartAtlasDbTransactionResponse;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.timestamp.TimestampRange;
+
+/**
+ * A timelock service decorator for introducing runtime validity checks on received timestamps.
+ */
+public final class TimestampCorroboratingTimelockService implements AutoDelegate_TimelockService {
```

Comment:
Copied from a now-dead comment

I don't think this should be a forwarding timelock service; at most it should be a TimestampService (which is the perfect interface for this)

In fact even on this own it's weird; with #3523 we'll never end up calling some of the methods and we never care about any thread safety because there's only one caller of timelock at a time

Reply:
For example, atop that PR, this change would be

```java
     @VisibleForTesting
     static Consumer<List<BatchElement<Integer, TimestampRange>>> consumer(TimestampService delegate) {
-        return batch -> {
-            long totalTimestamps = batch.stream().mapToLong(BatchElement::argument).reduce(0, Math::addExact);
-            long startInclusive = 0;
-            long endExclusive = 0;
-            for (BatchElement<Integer, TimestampRange> element : batch) {
-                int timestampsRequired = element.argument();
-                if (element.argument() <= endExclusive - startInclusive) {
-                    element.result().set(createExclusiveRange(startInclusive, startInclusive + timestampsRequired));
-                    startInclusive += timestampsRequired;
-                    totalTimestamps -= timestampsRequired;
+        return new Consumer<List<BatchElement<Integer, TimestampRange>>>() {
+            private long maxTimestampSeen = Long.MIN_VALUE;
+
+            @Override
+            public void accept(List<BatchElement<Integer, TimestampRange>> batch) {
+                long totalTimestamps = batch.stream().mapToLong(BatchElement::argument).reduce(0, Math::addExact);
+                long startInclusive = 0;
+                long endExclusive = 0;
+                for (BatchElement<Integer, TimestampRange> element : batch) {
+                    int timestampsRequired = element.argument();
+                    if (element.argument() <= endExclusive - startInclusive) {
+                        element.result().set(createExclusiveRange(startInclusive, startInclusive + timestampsRequired));
+                        startInclusive += timestampsRequired;
+                        totalTimestamps -= timestampsRequired;
+                    } else {
+                        TimestampRange requested = getFreshTimestamps(Ints.saturatedCast(totalTimestamps));
+                        startInclusive = requested.getLowerBound();
+                        endExclusive = Math.addExact(requested.getUpperBound(), 1);
+                        int toTake = Math.min(Ints.checkedCast(endExclusive - startInclusive), timestampsRequired);
+                        element.result().set(createExclusiveRange(startInclusive, startInclusive + toTake));
+                        startInclusive += toTake;
+                        totalTimestamps -= timestampsRequired;
+                    }
+                }
+            }
+
+            private TimestampRange getFreshTimestamps(int timestamps) {
+                TimestampRange range = getFreshTimestampsFromDelegate(delegate, timestamps);
+                if (maxTimestampSeen > range.getLowerBound()) {
+                    throw new RuntimeException("error");
                 } else {
-                    TimestampRange requested = getFreshTimestampsFromDelegate(
-                            delegate, Ints.saturatedCast(totalTimestamps));
-                    startInclusive = requested.getLowerBound();
-                    endExclusive = Math.addExact(requested.getUpperBound(), 1);
-                    int toTake = Math.min(Ints.checkedCast(endExclusive - startInclusive), timestampsRequired);
-                    element.result().set(createExclusiveRange(startInclusive, startInclusive + toTake));
-                    startInclusive += toTake;
-                    totalTimestamps -= timestampsRequired;
+                    maxTimestampSeen = range.getUpperBound();
                 }
             }
         };
```

Reply:
I think why we did it at this level was because we wanted to check the timestamps in 'compound ops' (lockImmutableTimestamp, startAtlasDbTransaction) which don't quite exist in TimestampService. Those methods make one RPC to timelock, so they don't actually call the underlying timestamp methods on the client side.

The compound ops aren't batched yet, though of course they could be batched in the future if we need to control timelock load

Reply:
is it worth doing the batching work in a follow up (i.e. after this PR) to avoid this undesirable middle state?

Reply:
It can be the case that the delegate gets called twice (or for that matter, O(num requests in batch) times), because of the contract of `TimestampService#getFreshTimestamps()` - returns at least 1, but not necessarily the number you ask for.

Reply:
I see! Thanks!